### ISSUE TITLE
Cannot login "ERROR: `mega-cli` terminated due to an error."

### ISSUE THREAD

**Issue by hellishvictor** [Hirevo/mega-cli-rs#1]
Hi, I used e-mail and pass in "mega-cli auth login" and I get:
```
`▱▱▱▱▱▱▱▱▰▰▰▰▰▰▰▰▰▰▱▱▱▱▱▱▱▱▱▱▱▱ logging in to MEGA...
▱▱▱▱▱▱▱▱▱▰▰▰▰▰▰▰▰▰▰▱▱▱▱▱▱▱▱▱▱▱ logging in to MEGA...
ERROR: `mega-cli` terminated due to an error.

  × could not log in to MEGA
  ╰─▶ could not get a meaningful response after maximum retries`
```
Cheers.

**Comment by Hirevo** [comment]
Hi, thanks for the bug report !

This is strange indeed, this error message happens when a request was unable to be successfully received or decoded and was retried as many times as the `max_retries` setting allowed.

A request gets retried if:
- the response timed out (the timeout duration is specified by the `timeout` setting)
- the request could not be made or the response could not be received (due to an I/O error or an HTTP error)
- the response's HTTP status code is not a successful one (not a 2XX status code)
- the response is an `EAGAIN` error from MEGA

A common case where this message can come up is if the HTTP request could not be made, like if there is no internet connectivity, or if a firewall blocked the request.

I'll work to add some optional logging into the mega crate to make these things more debuggable.  

I'll notify you right here when a new `mega-cli-rs` release with such logging will become available, so that we can hopefully fix this.

**Comment by hellishvictor** [comment]
Great, I'll look forward for the new release then. And btw, it would awesome if were added an option to store "default-config.toml" on the same folder where the program is (something like: "if not exist the config file here, create it on %APPDATA%\mega-cli-rs\config\"), so this way it can be used as portable on a USB pendrive or whatever.

**Comment by Hirevo** [comment]
> Great, I'll look forward for the new release then.

Optional logging is now available !  

To enable it, you need to set the `RUST_LOG` environment variable to `debug` before executing your command (most logs from `mega-rs` are at the `debug` level for now).

This should give out more information about what happened during the various request retries.

In theory, you shouldn't need to worry about the logs having some secret information as I took steps to redact as many personally identifying information from the output (things like user IDs, node IDs, email addresses, passwords, MFA tokens, and more).

But I'd encourage you to manually verify that this is the case before posting some logs here, in case I missed redacting something.

> And btw, it would awesome if were added an option to store "default-config.toml" on the same folder where the program is (something like: "if not exist the config file here, create it on %APPDATA%\mega-cli-rs\config"), so this way it can be used as portable on a USB pendrive or whatever.

I didn't consider that use-case of using it portably like that.  

It feels indeed like something akin to  a `--config-path` (`-c`) option is missing, which would allow relocating the config anywhere.  

Would you say such an option would satisfy your use-case ?

**Comment by hellishvictor** [comment]
> I didn't consider that use-case of using it portably like that.
> 
> It feels indeed like something akin to a `--config-path` (`-c`) option is missing, which would allow relocating the config anywhere.
> 
> Would you say such an option would satisfy your use-case ?

That's -c option is better than nothing,

**Comment by hellishvictor** [comment]
> It feels indeed like something akin to a `--config-path` (`-c`) option is missing, which would allow relocating the config anywhere.
> 
> Would you say such an option would satisfy your use-case ?

That "-c" option is better than nothing,

**Comment by hellishvictor** [comment]
Ok, finally today I got the time and this is what outputs the new v0.1.1:

```
▱▱▱▱▱▱▰▰▰▰▰▰▰▰▰▰▱▱▱▱▱▱▱▱▱▱▱▱▱▱ logging in to MEGA...
▱▱▱▱▱▱▱▰▰▰▰▰▰▰▰▰▰▱▱▱▱▱▱▱▱▱▱▱▱▱ logging in to MEGA...
▱▱▱▱▱▱▱▱▰▰▰▰▰▰▰▰▰▰▱▱▱▱▱▱▱▱▱▱▱▱ logging in to MEGA...
▱▱▱▱▱▱▱▱▱▰▰▰▰▰▰▰▰▰▰▱▱▱▱▱▱▱▱▱▱▱ logging in to MEGA...
▱▱▱▱▱▱▱▱▱▱▰▰▰▰▰▰▰▰▰▰▱▱▱▱▱▱▱▱▱▱ logging in to MEGA...
▱▱▱▱▱▱▱▱▱▱▱▰▰▰▰▰▰▰▰▰▰▱▱▱▱▱▱▱▱▱ logging in to MEGA...
▱▱▱▱▱▱▱▱▱▱▱▱▰▰▰▰▰▰▰▰▰▰▱▱▱▱▱▱▱▱ logging in to MEGA...
▱▱▱▱▱▱▱▱▱▱▱▱▱▰▰▰▰▰▰▰▰▰▰▱▱▱▱▱▱▱ logging in to MEGA...
←[2m2024-03-23T12:27:30.339241Z←[0m ←[31mERROR←[0m ←[2mmega::http::reqwest←[0m←[2m:←[0m `reqwest` er
ror when making MEGA request ←[3merror←[0m←[2m=←[0mreqwest::Error { kind: Request, url: Url { scheme
: "https", cannot_be_a_base: false, username: "", password: None, host: Some(Domain("g.api.mega.co.n
z")), port: None, path: "/cs", query: Some("id=0"), fragment: None }, source: hyper::Error(Connect,
Os { code: -2146762487, kind: Uncategorized, message: "Se procesó correctamente una cadena de certif
icados, pero termina en un certificado de raíz no compatible con el proveedor de confianza." }) }
←[2m2024-03-23T12:27:30.341241Z←[0m ←[31mERROR←[0m ←[2mmega::http::reqwest←[0m←[2m:←[0m maximum amou
nt of retries reached, cancelling MEGA request
▱▱▱▱▱▱▱▱▱▱▱▱▱▱▰▰▰▰▰▰▰▰▰▰▱▱▱▱▱▱ logging in to MEGA...
ERROR: `mega-cli` terminated due to an error.

  × could not log in to MEGA
  ╰─▶ could not get a meaningful response after maximum retries

```

**Comment by Hirevo** [comment]
Thanks for coming back with the logs.  

This error looks like TLS verification for the root certificate of `g.api.mega.co.nz` was rejected.  
Maybe your machine doesn't have this root certificate in its trust store.  

I'll start working on adding a `--no-check-certificate` flag, which would disable TLS certificate verification.  
Most HTTP/S clients have such an option, so it feels justified (`wget` has a `--no-check-certificate` flag and `curl` has a `-k|--insecure` flag).

**Comment by Hirevo** [comment]
I have now published `v0.1.2`, which contains the new `--no-check-certificate` global command-line flag.  

This option being global means it can be used for all subcommands, but it needs to be specified before the name of the subcommands.

Here is how you can use it:

```bash
# make sure `--no-check-certificate` comes before the subcommand name.
mega-cli --no-check-certificate auth login
```

This should fix the error you've seen.  
Feel free to comment here again if it did not, and I'll re-open the issue.

**Comment by hellishvictor** [comment]
**success: successfully logged in to MEGA !**

Thank you very much, man! Looking forward for the implementation of the -c command, and also, it would excelent if were added -ncc or whatever as the short version of the new command _--no-check-certificate_.

Cheers.
